### PR TITLE
Fix data corrupted issue when loading from UFS

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

A this fix a bug in corrupted data files.
Previously,  https://github.com/Alluxio/alluxio/pull/17497 attempt to solve this issue but only covers the case when creating a UFS reader
Notice we only handle block reader. So this may not fix `paged Block Reader` in 2.x 
### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

**Comment by jja725** [comment]
@yuzhu After this fix, we fix the issue for block reader. But this doesn't fix paged Block Reader & short circuit reader in 2.x, although these two is not recommended in 2.x. And there's still one test failed: `TierPromoteIntegrationTest.promoteByTierSwap`, which is also not recommended in 2.x. So do we want to spend more time to fix those?

**Comment by yuzhu** [comment]
Short circuit reader is not recommended for 2.x? That is news to me, why?

**Comment by jja725** [comment]
> Short circuit reader is not recommended for 2.x? That is news to me, why?

Maybe my information is not correct but I vaguely remember @apc999 want to deprecate this feature for a long time due to its complexity

**Comment by jja725** [comment]
@yuzhu @apc999  @JiamingMai I've been spending a day to try to trouble shooting the test(promoteByTierSwap) but get no luck. It's appreciated if any of you can help me take a look at the fix and see if I miss anything.

**Review by yuzhu — CHANGES_REQUESTED** [review]

File: core/common/src/main/java/alluxio/worker/block/io/BlockReader.java [hunk 1]
```diff
@@ -82,4 +82,17 @@ public BlockReader() {
...
+   * Abort the reader. This method should be called when an exception occurs during execution
+   * or when the operation is cancelled by the user.
+   */
+  public void abort() throws IOException {
+  }
+
+  /**
+   * Commit the reader in case there is any uncommitted block cached from UFS.
+   */
+  public void commit() throws IOException {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/worker/block/io/BlockReader.java:96]: I don't see why need a separate `commit()`, let's discuss offline

**Review by apc999 — CHANGES_REQUESTED** [review]
Comments left

File: tests/src/test/java/alluxio/server/tieredstore/TierPromoteIntegrationTest.java [hunk 2]
```diff
@@ -98,6 +99,7 @@ public TierPromoteIntegrationTest(String shortCircuitEnabled) {
   @LocalAlluxioClusterResource.Config(confParams = {
       PropertyKey.Name.USER_FILE_WRITE_TYPE_DEFAULT, "MUST_CACHE"})
   @Test
+  @Ignore
```
Comment (apc999) [inline tests/src/test/java/alluxio/server/tieredstore/TierPromoteIntegrationTest.java:102]:
```suggestion
  @Ignore("TierPromote is not recommended")
```

File: core/common/src/main/java/alluxio/worker/block/io/BlockReader.java (hunk 1 above)
Comment (jja725) [reply core/common/src/main/java/alluxio/worker/block/io/BlockReader.java:96]: Yes we can combine these two functions but I feel it's more clear to separate the responsibility of close resources & commit block.

**Review by apc999 — APPROVED** [review]
thanks for the fix!

File: core/server/worker/src/main/java/alluxio/worker/block/CacheRequestManager.java [hunk 3]
```diff
@@ -281,10 +283,26 @@ private CacheResult cacheBlockFromUfs(long blockId, long blockSize,
         reader.read(offset, bufferSize);
         offset += bufferSize;
       }
+      if (isTempBlockMetaPresent(blockId)) {
+        mBlockWorker.commitBlock(Sessions.CACHE_UFS_SESSION_ID, blockId, false);
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed to async cache block {} from UFS on reading the block: {}", blockId, e);
```
[the remark this answers is not in the document]
Comment (jja725) [reply core/server/worker/src/main/java/alluxio/worker/block/CacheRequestManager.java]: done

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.10 please

**Pull request merged**